src/Solvers/conjugate_gradient_solver.jl

```diff
@@ -71,14 +77,24 @@ Arguments
 * `preconditioner`: Object for which `precondition!(z, preconditioner, r, args...)` computes `z = P * r`,
                     where `r` is the residual. Typically `P` is approximately `A⁻¹`.
 
+* `enforce_gauge_condition!`: Function with signature `enforce_gauge_condition!(x, r)` that
+                              enforces a gauge condition on the solution `x` and residual `r`.
+                              This is useful for problems where the solution is not unique, such as
+                              the Poisson equation with purely Neumann boundary conditions. 
+                              The function is called at the end of each iteration of a conjugate 
+                              gradient iteration to ensure that the solution remains consistent 
+                              with the gauge condition.
+                              The default is `no_gauge_enforcement!`, which does not enforce a gauge condition.
+
 See [`solve!`](@ref) for more information about the preconditioned conjugate-gradient algorithm.
 """
 function ConjugateGradientSolver(linear_operation;
                                  template_field::AbstractField,
                                  maxiter = prod(size(template_field)),
                                  reltol = sqrt(eps(eltype(template_field.grid))),
                                  abstol = 0,
-                                 preconditioner = nothing)
+                                 preconditioner = nothing, 
+                                 enforce_gauge_condition! = no_gauge_enforcement!)
```

Comment:
I think this should follow the same notation as `ConjugateGradientPoissonSolver`, no?
```suggestion
                                 gauge_condition = no_gauge_enforcement!)
```

Reply:
what if we do it the other way round ie: change https://github.com/CliMA/Oceananigans.jl/blob/fe97955146016c6659ce02ffa420b2f272f24df7/src/Solvers/conjugate_gradient_poisson_solver.jl#L86 to `enforce_gauge_condition!` instead? In my mind I was hoping this argument to be a verb to imply that it is a function-like object, rather than a noun such as `gauge_condition` which to me implies something more like a number. But of course I am happy to do it the other way as well.

Reply:
@glwagner is the expert on notation here, so I'm curious about what he has to say here. But my opinion is that if the keyword is `enforce_gauge_condition`, a user would expect it to receive a boolean. Like, "yes enforce gauge condition", or "no don't enforce a gauge condition". When you need pass a function with a specific gauge condition as a kwarg, then I think it's more intuitive for the kwarg to be a noun. Does that make sense?

Of course, that's talking from a user perspective, who most of the time won't read the code. In the code, the kwarg will "become" the function and you'll eventually call `gauge_condition(x, r)` (or `enforce_gauge_condition!(x, r)`). In that context the latter makes more sense.

Given that more people will use the function without reading its source code, I feel like it's better to use `gauge_condition`. But I also don't feel strongly about it.

Maybe `gauge_condition_function` would be a good compromise... ? idk

Reply:
I think `enforce_gauge_condition!` is a verb so it can be expected to be a function / callable. (the noun version, which might be a Boolean, might be something like `gauge_enforcement`)